zebROS_ws/src/behaviors/src/2023_fourber_server.cpp

```diff
@@ -0,0 +1,383 @@
+#include <ros/ros.h>
+#include <actionlib/server/simple_action_server.h>
+#include <behavior_actions/Fourber2023Action.h>
+#include <ddynamic_reconfigure/ddynamic_reconfigure.h>
+#include <std_msgs/Float64.h>
+#include <controllers_2023_msgs/FourBarSrv.h>
+#include <iostream>
+#include <talon_state_msgs/TalonState.h>
+#include <atomic>
+
+#define FourberINFO(x) ROS_INFO_STREAM("2023_fourber_server : " << x)
+#define FourberWARN(x) ROS_WARN_STREAM("2023_fourber_server : " << x)
+#define FourberERR(x) ROS_ERROR_STREAM("2023_fourber_server : " << x)
+
+typedef behavior_actions::Fourber2023Goal fourber_ns;
+
+constexpr std::array mode_to_string = {"INTAKE", "LOW_NODE", "MIDDLE_NODE", "HIGH_NODE"};
+constexpr std::array piece_to_string = {"CUBE", "VERTICAL_CONE", "BASE_TOWARDS_US_CONE", "BASE_AWAY_US_CONE"};
+
+/*
+# Fourber = 
+uint8 CUBE=0
+uint8 VERTICAL_CONE=1
+uint8 BASE_TOWARDS_US_CONE=2 
+uint8 BASE_AWAY_US_CONE=3
+uint8 piece
+
+uint8 INTAKE_LOW=0 # should be just two places that we need to call saftey with: when intaking and when going above a certain height
+uint8 HIGH=1 
+uint8 saftey_position # looks up config value on where to go based on this, ignored if saftey is false
+
+---
+#result
+bool success # if we got there
+---
+#feedback
+bool success # if we got there
+*/
+
+template <class T>
+void load_param_helper(const ros::NodeHandle &nh, std::string name, T &result, T default_val) {
+    if (!nh.getParam(name, result))
+    {
+      FourberERR("Could not find" << name << ", defaulting to " << default_val);
+      result = default_val;
+    }
+}
+
+enum SafteyState {
+  NONE,
+  SAFTEY_HIGH,
+  SAFTEY_LOW
+};
+
+class FourberAction2023 {
+
+protected:
+
+  ros::NodeHandle nh_;
+  ros::NodeHandle nh_params_;
+  actionlib::SimpleActionServer<behavior_actions::Fourber2023Action> as_;
+  
+  // class memebers for the publishFailure function, reduces code duplication
+  behavior_actions::Fourber2023Feedback feedback_;
+  behavior_actions::Fourber2023Result result_;
+
+  double safety_high_distance_;
+  double safety_low_distance_;
+  bool safety_high_below_;
+  bool safety_low_below_;
+
+  ros::ServiceClient fourbar_srv_;
+  std::string action_name_;
+
+  ddynamic_reconfigure::DDynamicReconfigure ddr_;
+  ros::Subscriber fourbar_offset_sub_;
+  ros::Subscriber talon_states_sub_;
+  std::map<int, std::map<int, std::pair<double, bool>>> game_piece_lookup_;
```

Comment:
This is complex enough it might make sense to make it a bit cleaner

That is, a struct for the inner std::pair so the double and bool have descriptive names.
And instead of a map<int<map<int, blah>>>, make a struct with two different enum types (game piece, mode) and use that as the key.

My immediate worry here is all of the nested lookups with zero error checking on them.

This should probably apply to the elevator code as well, meaning it might make sense to put those enums and a typedef for the map into a shared header file. Then each .cpp can declare their own instance of it.
Or even better, if the code to load the config values is the same, pull those functions out into a separate cpp file and use that one file in both servers. That way the code isn't duplicated, meaning fixes to one automatically make it into the other.